lib/msal-common/src/cache/CacheManager.ts

```diff
@@ -261,60 +291,78 @@ export abstract class CacheManager implements ICacheManager {
 
-    /**
-     * retrieve accounts matching all provided filters; if no filter is set, get all accounts
-     * not checking for casing as keys are all generated in lower case, remember to convert to lower case if object properties are compared
-     * @param homeAccountId
-     * @param environment
-     * @param realm
-     */
-    private getAccountsFilteredByInternal(
-        homeAccountId?: string,
-        environment?: string,
-        realm?: string,
-        nativeAccountId?: string,
-    ): AccountCache {
-        const allCacheKeys = this.getKeys();
-        const matchingAccounts: AccountCache = {};
+        allAccountKeys.forEach((cacheKey) => {
+            if (!this.isAccountKey(cacheKey, accountFilter.homeAccountId, accountFilter.realm)) {
+                // Don't parse value if the key doesn't match the account filters
+                return;
+            }
 
-        allCacheKeys.forEach((cacheKey) => {
             const entity: AccountEntity | null = this.getAccount(cacheKey);
 
             if (!entity) {
                 return;
             }
 
-            if (!!homeAccountId && !this.matchHomeAccountId(entity, homeAccountId)) {
+            if (!!accountFilter.homeAccountId && !this.matchHomeAccountId(entity, accountFilter.homeAccountId)) {
                 return;
             }
 
-            if (!!environment && !this.matchEnvironment(entity, environment)) {
+            if (!!accountFilter.localAccountId && !this.matchLocalAccountId(entity, accountFilter.localAccountId)) {
                 return;
             }
 
-            if (!!realm && !this.matchRealm(entity, realm)) {
+            if (!!accountFilter.username && !this.matchUsername(entity, accountFilter.username)) {
                 return;
             }
 
-            if (!!nativeAccountId && !this.matchNativeAccountId(entity, nativeAccountId)) {
+            if (!!accountFilter.environment && !this.matchEnvironment(entity, accountFilter.environment)) {
                 return;
             }
 
-            matchingAccounts[cacheKey] = entity;
+            if (!!accountFilter.realm && !this.matchRealm(entity, accountFilter.realm)) {
+                return;
+            }
+
+            if (!!accountFilter.nativeAccountId && !this.matchNativeAccountId(entity, accountFilter.nativeAccountId)) {
+                return;
+            }
+
+            matchingAccounts.push(entity);
         });
 
         return matchingAccounts;
     }
 
+    /**
+     * Returns true if the given key matches our account key schema. Also matches homeAccountId and/or tenantId if provided
+     * @param key 
+     * @param homeAccountId 
+     * @param tenantId 
+     * @returns 
+     */
+    isAccountKey(key: string, homeAccountId?: string, tenantId?: string): boolean {
+        if (key.split(Separators.CACHE_KEY_SEPARATOR).length < 3) {
+            // Account cache keys contain 3 items separated by '-' (each item may also contain '-')
```

Comment:
Sorry, the separator is '-', but each item can also contain '-' in it?

Reply:
Yes...an unfortunate decision